src/hibernate-resume/hibernate-resume.c

```diff
@@ -1,58 +1,76 @@
 /* SPDX-License-Identifier: LGPL-2.1-or-later */
 
 #include <errno.h>
-#include <stdio.h>
 #include <sys/stat.h>
 
-#include "alloc-util.h"
 #include "devnum-util.h"
 #include "fileio.h"
 #include "initrd-util.h"
 #include "log.h"
+#include "main-func.h"
+#include "parse-util.h"
+#include "static-destruct.h"
+#include "string-util.h"
 
-int main(int argc, char *argv[]) {
+static char *arg_resume_device = NULL;
+static uint64_t arg_resume_offset = 0;
+
+STATIC_DESTRUCTOR_REGISTER(arg_resume_device, unsetp);
+
+static int run(int argc, char *argv[]) {
         struct stat st;
-        const char *device;
         int r;
 
-        if (argc != 2) {
-                log_error("This program expects one argument.");
-                return EXIT_FAILURE;
-        }
-
         log_setup();
 
+        if (argc != 3)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "This program expects two arguments.");
+
         umask(0022);
 
-        /* Refuse to run unless we are in an initrd() */
         if (!in_initrd())
-                return EXIT_SUCCESS;
+                return 0;
 
-        device = argv[1];
+        arg_resume_device = argv[1];
 
-        if (stat(device, &st) < 0) {
-                log_error_errno(errno, "Failed to stat '%s': %m", device);
-                return EXIT_FAILURE;
-        }
+        r = safe_atou64(argv[2], &arg_resume_offset);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse resume offset %s: %m", argv[2]);
 
-        if (!S_ISBLK(st.st_mode)) {
-                log_error("Resume device '%s' is not a block device.", device);
-                return EXIT_FAILURE;
-        }
+        if (stat(arg_resume_device, &st) < 0)
+                return log_error_errno(errno, "Failed to stat resume device '%s': %m", arg_resume_device);
 
-        r = write_string_file("/sys/power/resume", FORMAT_DEVNUM(st.st_rdev), WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (!S_ISBLK(st.st_mode))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Resume device '%s' is not a block device.", arg_resume_device);
+
+        r = write_string_filef("/sys/power/resume_offset", WRITE_STRING_FILE_DISABLE_BUFFER, "%" PRIu64, arg_resume_offset);
```

Comment:
so here you write the offset first, the device second. in the sleep code the other way round. Wouldn't writing the offset first, and the device second be the safer choice always?

also, it feels as if we should have a common helper `int write_sleep_device(dev_t device, uint64_t offset)` or so?